crates/kubelet/src/resources/device_plugin_manager/node_patcher.rs

```diff
@@ -0,0 +1,260 @@
+use super::{DeviceMap, HEALTHY};
+use k8s_openapi::api::core::v1::Node;
+use kube::api::{Api, PatchParams};
+use std::sync::Arc;
+use tokio::sync::{broadcast, RwLock};
+use tracing::{debug, error};
+
+/// NodePatcher updates the Node status with the latest device information.
+#[derive(Clone)]
+pub struct NodeStatusPatcher {
+    node_name: String,
+    devices: Arc<RwLock<DeviceMap>>,
+    // Broadcast sender so clonable
+    update_node_status_sender: broadcast::Sender<()>,
+    client: kube::Client,
+}
+
+impl NodeStatusPatcher {
+    pub fn new(
+        node_name: &str,
+        devices: Arc<RwLock<DeviceMap>>,
+        update_node_status_sender: broadcast::Sender<()>,
+        client: kube::Client,
+    ) -> Self {
+        NodeStatusPatcher {
+            node_name: node_name.to_string(),
+            devices,
+            update_node_status_sender,
+            client,
+        }
+    }
+
+    // TODO: Decide whether `NodePatcher` should do `remove` patches when there are no
+    // devices under a resource. When a device plugin drops, the `DeviceManager` clears
+    // out the resource's device map. Currently, this just sets the resource's
+    // `allocatable` and `capacity` count to 0, which appears to be the same implementation
+    // in Kubernetes.
+    async fn get_node_status_patch(&self) -> json_patch::Patch {
```

Comment:
this should possibly be called by the `update_status` function that is called every 10 seconds in kubelet. Otherwise, the resources' availability is only updated on allocate calls, meaning that if a Pod that requested a device plugin resource terminates, the amount of that resource available (`NodeStatus.allocatable`) is one too few until the next allocate call (next Pod requesting a resource is scheduled). This means that a pod will not be scheduled if the `NodeStatus.allocatable` for that resource is 0. The only way it would be updated to 1 is if a different resource is requested and triggers the update.
Instead, a reference to the nodepatcher could be added to `start_node_updater` in `kubelet.rs`

Reply:
I am fine with either option here. It sounds like leaving it as-is is not the best option

Reply:
To clarify, the TODO documentation above the `get_node_status_patch` function is referring to a different consideration than my comment here. The TODO is to decide how fully to clean up resources that no longer have a device plugin. This comment is to decide when to call `update_allocated_devices` and then patch the node. Currently, `update_allocated_devices` (which checks what pods are currently using resources) is only called during an `allocate` call.

Reply:
I did some thinking here and i believe the current implementation is correct as is. I confused myself and thought for a second that `capacity` and `allocatable` change with regards to pods using the resource, but they dont. they only reflect the quantity and health of the device. I also confirmed this with testing on k8s kubelet locally. disregard my first comment in this thread. The only way "`NodeStatus.allocatable` for that resource is 0" is if the device plugin reports the device unhealthy or if the device plugin goes offline